File: day2/day2.rb
lines = File.read('input.txt').split("\n")

passwords_matching = lines.count do |line|
  /(?<min>\d*)-(?<max>\d*) (?<letter>\w): (?<password>.*)/ =~ line

  password.count(letter).between?(min.to_i, max.to_i)
end

puts '----part1----'
puts passwords_matching

puts '----part2----'

passwords_matching_2 = lines.count do |line|
  /(?<min>\d*)-(?<max>\d*) (?<letter>\w): (?<password>.*)/ =~ line

  (password[min.to_i - 1] == letter && password[max.to_i - 1] != letter) ||
    (password[min.to_i - 1] != letter && password[max.to_i - 1] == letter)
end

puts passwords_matching_2

File: 2020/day2/day2.go
package main

import (
    "fmt"
    "io/ioutil"
    "strconv"
    "strings"
)

func main() {
    content, err := ioutil.ReadFile("input.txt")

    if err != nil {
      panic(err)
    }

    lines := strings.Split(string(content), "\n")

    part1(lines)
    part2(lines)
}

func part1(lines []string) {
    matching := 0

    for i, l := range lines {
      if i == len(lines) - 1 {
        break
      }

      bags := strings.Split(l, ": ")

      password := bags[1]

      sub_bags := strings.Split(bags[0], " ")
      min_max := strings.Split(sub_bags[0], "-")
      min, _ := strconv.Atoi(min_max[0])
      max, _ := strconv.Atoi(min_max[1])
      letter := sub_bags[1]

      occurrences := strings.Count(password, letter)

      if occurrences >= min && occurrences <= max {
        matching++
      }
    }

    fmt.Println(matching)
}

func part2(lines []string) {
    matching := 0

    for i, l := range lines {
      if i == len(lines) - 1 {
        break
      }

      bags := strings.Split(l, ": ")

      password := bags[1]

      sub_bags := strings.Split(bags[0], " ")
      min_max := strings.Split(sub_bags[0], "-")
      min, _ := strconv.Atoi(min_max[0])
      max, _ := strconv.Atoi(min_max[1])
      letter := sub_bags[1]

      first_candidate := string(password[min - 1])
      second_candidate := string(password[max - 1])
      if (first_candidate == letter && second_candidate != letter) ||
           (first_candidate != letter && second_candidate == letter) {
        matching++
      }
    }

    fmt.Println(matching)
}
